### ISSUE TITLE
Why is TF 2.2 much slower than TF 2.0?

### ISSUE THREAD

**Issue by OverLordGoldDragon** [tensorflow/tensorflow#39665]
Followup to [original](https://github.com/tensorflow/tensorflow/issues/33487); I've benched various configurations per code [here](https://stackoverflow.com/questions/58441514/why-is-tensorflow-2-much-slower-than-tensorflow-1#answer-58653636), and found "Large model on Large data" to perform much worse in TF 2.2 than in TF 2.0 or TF 1.14.0. Code for that particular case, and result plots, below.

Why is this the case? To me, this is a clear dealbreaker for upgrading.

<hr>

**Test plots**:

![image](https://user-images.githubusercontent.com/16495490/82284722-bba71200-99aa-11ea-9877-933e45d5d945.png)

Per above, Graph and Eager are **1.56x** and **1.97x** slower than their TF1 counterparts, respectively. 

<hr>

<details>
  <summary><b>Test code</b></summary>

```python
import numpy as np
import tensorflow as tf
import random
import matplotlib.pyplot as plt
from termcolor import cprint
from time import time

from tensorflow.keras.layers import Input, Dense, Conv1D
from tensorflow.keras.layers import Dropout, GlobalAveragePooling1D
from tensorflow.keras.models import Model
from tensorflow.keras.optimizers import Adam
import tensorflow.keras.backend as K

# tf.compat.v1.disable_eager_execution()

###############################################################################
def reset_seeds(reset_graph_with_backend=None, verbose=1):
    if reset_graph_with_backend is not None:
        K = reset_graph_with_backend
        K.clear_session()
        tf.compat.v1.reset_default_graph()
        if verbose:
            print("KERAS AND TENSORFLOW GRAPHS RESET")

    np.random.seed(1)
    random.seed(2)
    tf.compat.v1.set_random_seed(3)
    if verbose:
        print("RANDOM SEEDS RESET")

print("TF version: {}".format(tf.__version__))
reset_seeds()

def timeit(func, iterations, *args, _verbose=0, **kwargs):
    times = []
    t0 = time()
    for _ in range(iterations):
        t1 = time()
        func(*args, **kwargs)
        times.append(time() - t1)
        print(end='.'*int(_verbose))
    print("Time/iter: %.4f sec" % ((time() - t0) / iterations))
    plt.stem(times)
    plt.ylim(0, 15)

###############################################################################
def make_model_large(batch_shape):
    ipt   = Input(batch_shape=batch_shape)
    x     = Conv1D(64,  400, strides=4, padding='valid')(ipt)
    x     = Conv1D(128, 200, strides=1, padding='valid')(x)
    for _ in range(40):
        x = Conv1D(256,  12, strides=1, padding='same')(x)
    x     = Conv1D(512,  20, strides=2, padding='valid')(x)
    x     = Conv1D(1028, 10, strides=2, padding='valid')(x)
    x     = Conv1D(256,   1, strides=1, padding='valid')(x)
    x     = GlobalAveragePooling1D()(x)
    x     = Dense(256, activation='relu')(x)
    x     = Dropout(0.5)(x)
    x     = Dense(128, activation='relu')(x)
    x     = Dense(64,  activation='relu')(x)
    out   = Dense(1,   activation='sigmoid')(x)
    model = Model(ipt, out)
    model.compile(Adam(lr=1e-4), 'binary_crossentropy')
    return model

def make_data(batch_shape):
    return np.random.randn(*batch_shape), \
           np.random.randint(0, 2, (batch_shape[0], 1))

def test_all():
    for model_fn, model_name, iters in zip(make_model_fns, model_names,
                                           iterations):
        for batch_shape, shape_name in zip(batch_shapes, shape_names):
            reset_seeds(reset_graph_with_backend=K)
            data = make_data(batch_shape)
            model = model_fn(batch_shape)

            model.train_on_batch(*data)
            timeit(model.train_on_batch, iters, *data, _verbose=1)
            
            cprint(">> {}, {} done <<\n".format(model_name, shape_name), 'blue')
            del model
###############################################################################

batch_shape_large  = (32, 14000, 30)
batch_shape = batch_shape_large
model_fn = make_model_large

batch_shapes = batch_shape_large,
make_model_fns = make_model_large,
iterations = [200]
shape_names = ["Large data"]
model_names = ["Large model"]

test_all()
```

</details>

**Comment by amahendrakar** [comment]
@OverLordGoldDragon,
Running the given code on Google Colab takes around 400 seconds per iteration.

Could you please provide a minimal code sample to reproduce the issue reported here. Thanks!

**Comment by mihaimaruseac** [comment]
Can you compare tf-nightly? So we can test master code

**Comment by OverLordGoldDragon** [comment]
@amahendrakar Did you use the GPU backend? Runs fine for me

**Comment by OverLordGoldDragon** [comment]
@mihaimaruseac tf-nightly plots for 100 iterations (200 is lengthy); the Eager numbers are nicer per being biased by the initial 'warmup' - but strangely enough, Graph is worse now, though this might be due to me taking less time between successive tests this time around (I did Eager first). Either way, still much slower than TF 2.0.

![image](https://user-images.githubusercontent.com/16495490/82385682-b518a900-9a43-11ea-9491-be4585623a02.png)

**Comment by amahendrakar** [comment]
Was able to reproduce the issue. Please check the below plots and linked gist for the respective TensorFlow versions.
- [TF v1.14](https://colab.research.google.com/gist/amahendrakar/ade7deaed275faae234cb39c0d5aceb0/39665-1-14.ipynb#scrollTo=LSQqiFqAhgoR) and [TF v1.15](https://colab.research.google.com/gist/amahendrakar/998c99bd90902986e596c3953b861046/39665-1-15.ipynb#scrollTo=LSQqiFqAhgoR)
![1 x](https://user-images.githubusercontent.com/57165142/82440116-7291b380-9ab9-11ea-851c-7be010479c1d.png)

- [TF v2.0 eager mode](https://colab.research.google.com/gist/amahendrakar/0a152d6ebe5f840592c57c3c5c3eea22/39665-2-0.ipynb#scrollTo=tvE7GOv8ZhJ6&line=1&uniqifier=1) and [TF v2.0 graph mode](https://colab.research.google.com/gist/amahendrakar/70e5667b5061af28af14c28a65e3bd75/39665-2-0-graph.ipynb#scrollTo=tvE7GOv8ZhJ6&line=95&uniqifier=1)
![2 0](https://user-images.githubusercontent.com/57165142/82440294-b684b880-9ab9-11ea-848a-25f42333fa9b.png)

- [TF v2.1 eager mode](https://colab.research.google.com/gist/amahendrakar/aeb204c3cb4071c423b197f2bbc4baa0/39665-2-1.ipynb#scrollTo=LSQqiFqAhgoR) and [TF v2.1 graph mode](https://colab.research.google.com/gist/amahendrakar/139a8bb3b52a43746bb6a9ec67311ef4/39665-2-1-graph.ipynb#scrollTo=LSQqiFqAhgoR)
![2 1](https://user-images.githubusercontent.com/57165142/82440481-02376200-9aba-11ea-9076-fb9a45ad105a.png)

- [TF v2.2 eager mode](https://colab.research.google.com/gist/amahendrakar/04d2152837786cc6b689241fd6b72dba/39665-2-2.ipynb#scrollTo=LSQqiFqAhgoR) and [TF v2.2 graph mode](https://colab.research.google.com/gist/amahendrakar/158879ba379abe8df65292f3360bac07/39665-2-2-graph.ipynb#scrollTo=LSQqiFqAhgoR)
![2 2](https://user-images.githubusercontent.com/57165142/82440654-52162900-9aba-11ea-8ab9-6a0d1016fa7a.png)

- [TF nightly eager mode](https://colab.research.google.com/gist/amahendrakar/48357c72d4bfa45bf87b0776733dc42a/39665-tf-nightly.ipynb#scrollTo=LSQqiFqAhgoR) and [TF nightly graph mode](https://colab.research.google.com/gist/amahendrakar/c4f351a921d3c547d0f5a20e2fec1de7/39665-nightly-graph.ipynb#scrollTo=LSQqiFqAhgoR)
![2 3](https://user-images.githubusercontent.com/57165142/82440764-7e31aa00-9aba-11ea-8e91-46509015eadc.png)

Thanks!

**Comment by OverLordGoldDragon** [comment]
@amahendrakar Thanks for the encompassing testing. TF2.0-Graph is unusually fast per these benchmarks, and TF1.14.0 is unusually slow; I wonder if the difference in GPU's is responsible - my benchmarks are based on my local GTX 1070. I'll look into it.

**Comment by OverLordGoldDragon** [comment]
Regarding my plots, turns out my GPU's been overheating lately, thus the rising stem plots.

However, the problem isn't fully solved; [followup](https://github.com/tensorflow/tensorflow/issues/42429).

**Comment by mohantym** [comment]
Hi @OverLordGoldDragon ! 
We are checking to see whether you still need help in this issue . Have you tried your latest version (TF 2.7 ) yet?

**Comment by OverLordGoldDragon** [comment]
Been resolved since 2.3, but also haven't checked since - at any rate I'll mark it resolved.

**Comment by mohantym** [comment]
Are you satisfied with the resolution of your issue?
[Yes](https://goo.gl/forms/Oe0tEvODFRoI2gJF3)
[No](https://goo.gl/forms/fUjzOfrtkFbrOT8d2)
